dask/dataframe/core.py

```diff
@@ -4672,15 +4673,82 @@ def set_index(
         """
+
+        # Deal with bool params
         if inplace:
             raise NotImplementedError("The inplace= keyword is not supported")
         pre_sorted = sorted
         del sorted
 
-        if isinstance(other, Series) and other._name == self.index._name:
-            # Index is equal to current index, no-op
-            return self
+        # Check other can be translated to column name or column object, possibly flattening it
+        if not isinstance(other, str):
+            if isinstance(other, Sequence):
+                # Accept ["a"], but not [["a"]]
+                if len(other) == 1 and (
+                    isinstance(other[0], str) or not isinstance(other[0], Sequence)
+                ):
+                    other = other[0]
+                else:
+                    raise NotImplementedError(
+                        "Dask dataframe does not yet support multi-indexes.\n"
+                        f"You tried to index with this index: {other}\n"
+                        "Indexes must be single columns only."
+                    )
+
+            # Can also be a frame
+            elif isinstance(other, DataFrame):
+                raise NotImplementedError(
+                    "Dask dataframe does not yet support multi-indexes.\n"
+                    f"You tried to index with a frame with these columns: {list(other.columns)}\n"
+                    "Indexes must be single columns only."
+                )
 
+        # If already a series
+        index, columns = self.index, self.columns
+        if is_series := isinstance(other, Series):
+            # If it's already the index, there's nothing to do
+            if other._name == index._name:
+                return self
+            # Otherwise, check length matches when other isn't one of the data columns
+            s = other
+            if not any(
+                s._name == col_series._name for _, col_series in self.items()
+            ) and (  # for cases with other=df.column
+                len_index := len(s)
+            ) != (
+                len_self := len(self)
+            ):
+                raise ValueError(
+                    f"Length mismatch: series to become index has {len_index} rows, but data have {len_self} rows"
+                )
+
+        # If the name of a column/index
+        else:
+            # If the index, no additional checks
+            if (
+                other == index.name
+            ):  # here it's name that should be checked, instead of _name
+                return self
+            # If a missing column, KeyError
+            if other not in columns:
+                raise KeyError(
+                    f"Data has no column '{other}': use either index '{index.name}' or any column of {list(columns)}"
+                )
+            s = self[other]
+
+        # Ensure there are no nulls
+        if s.isna().any().compute():
```

Comment:
If we think it's important to compute nulls, then I think we should compute them when we do min, max, and len. That will allow us to only read in the data once. If that means that we miss the case where the index is sorted and the divisions are passed, I think that's ok.

Reply:
Not sure I'm following you. Currently it is when the code enters the "divisions-figuring-out-part" in `shuffle.py` that things fail, because it's not robust to nulls. So I'm doing this previously to prevent that happens.

Of course, when that part is robust to allow for nulls (and also Pandas extension dtypes that use `pd.NA`), then we'll be able to remove this test safely.

Reply:
Right so could we fail within that method when we encounter a nan? We could wrap this https://github.com/dask/dask/blob/138831b8f6cdf2e7875d186d715ba87edd041936/dask/dataframe/shuffle.py#L42 in a try except or we could drill down into https://github.com/dask/dask/blob/138831b8f6cdf2e7875d186d715ba87edd041936/dask/dataframe/partitionquantiles.py#L443 and raise from somewhere within that function.

Reply:
Wrapped the `compute` line into a `try-except`. Interestingly, that only fails when the column is non-numeric (`nan` is fine), so I was more flexible regarding what to accept in `from_pandas`. Now errors only happen when: 1) the index has some null, 2) it is non-numeric.

Even so, I still think that for numeric cases nulls shouldn't be really accepted as part of the index. Correct me if I'm wrong, but the main purpose of Dask's index is to know in what partition to look for a particular value (that is, for `loc`), right? 

However, given that comparisons with null values don't always yield `True` (`np.nan == np.nan` yields `False`!!!), and that you can't really tell whether `nan` is greater than or less than some non-null value, I think allowing for that can only bring trouble.

Reply:
Update: I slightly improved this so that `nulls_presence` is computed along with `sizes, mins, maxes`, etc. When an exception is raised, there are nulls and the series `dtype` is non-numeric, we inform the user about those nulls being likely the cause of the problem. Hope you find this reasonable.